### PR TITLE
Generate specs from TGI openapi.json

### PR DESCRIPTION
**Pull request by Wauplin**
From an idea mentioned by @OlivierDehaene and @drbh in https://github.com/huggingface/huggingface.js/pull/579 and [slack thread](https://huggingface.slack.com/archives/C05CFK1HM0T/p1711360022125399) (internal).

This PR adds a script  `inference-tgi-import.ts` to generate the `text-generation` and `chat_completion` specs from the auto-generated TGI [specifications](https://huggingface.github.io/text-generation-inference/). The goal is to keep in sync TGI improvements with @huggingface/tasks and therefore have a consistent "single source of truth". The converted specs are then compatible with our tooling to generate JS/Python code.

This PR changes quite a lot of naming in the generated JS/Python types. Luckily I don't think they were yet used in the JS ecosystem so better to do that now than later.

I also opened https://github.com/huggingface/huggingface_hub/pull/2237 to include these changes in `huggingface_hub`.

**TODO:**
- [ ] fix lint errors. How to deal with a parsed json to avoid using `any`? 
- [ ] CI workflow to open a PR each time TGI is updated? => can be done in a future PR

### CODE REVIEW

File: packages/tasks/src/tasks/text-generation/inference.ts [hunk 1]
```diff
@@ -5,204 +5,94 @@
...
-export interface TextGenerationParameters {
-	/**
-	 * The number of sampling queries to run. Only the best one (in terms of total logprob) will
-	 * be returned.
-	 */
+export interface TextGenerationInputGenerateParameters {
 	best_of?: number;
-	/**
-	 * Whether or not to output decoder input details
-	 */
```
Comment (SBrandeis) [inline packages/tasks/src/tasks/text-generation/inference.ts:39]: How would we proceed to generate JSDocs like this?
I think this is very useful for downstream consumers

File: packages/tasks/src/tasks/text-generation/inference.ts [hunk 2]
```diff
@@ -5,204 +5,94 @@
...
+	 *
+	 * JSON Schema is a declarative language that allows to annotate JSON documents
+	 * with types and descriptions.
 	 */
-	watermark?: boolean;
+	value: unknown;
 	[property: string]: unknown;
 }
 
+export type Type = "json" | "regex";
```
Comment (SBrandeis) [inline packages/tasks/src/tasks/text-generation/inference.ts:50]: This new type's name is a bit too generic IMO

File: packages/tasks/src/tasks/text-generation/spec/input.json [hunk 3]
```diff
@@ -1,94 +1,204 @@
...
+			"properties": {
+				"inputs": {
+					"type": "string",
+					"example": "My name is Olivier and I"
+				},
+				"parameters": {
+					"$ref": "#/$defs/TextGenerationInputGenerateParameters"
+				}
+			}
+		},
```
Comment (SBrandeis) [inline packages/tasks/src/tasks/text-generation/spec/input.json:30]: This is a re-declaration of the "root type" it seems - should not be needed right?

File: packages/tasks/src/tasks/text-generation/spec/input.json [hunk 4]
```diff
@@ -1,94 +1,204 @@
...
 			"type": "object",
 			"properties": {
 				"best_of": {
 					"type": "integer",
-					"description": "The number of sampling queries to run. Only the best one (in terms of total logprob) will be returned."
+					"default": "null",
+					"example": 1,
+					"nullable": true,
+					"minimum": 0,
+					"exclusiveMinimum": 0
```
Comment (SBrandeis) [inline packages/tasks/src/tasks/text-generation/spec/input.json:40]: 😄 You got your input validation info in the end

File: packages/tasks/src/tasks/text-generation/spec/output.json [hunk 5]
```diff
@@ -1,165 +1,197 @@
...
-				{ "const": "eos_token", "description": "eos_token: The model generated an end-of-sentence (EOS) token" },
-				{
-					"const": "stop_sequence",
-					"description": "stop_sequence: One of the sequence in stop_sequences was generated"
-				}
-			]
-		},
-		"PrefillToken": {
-			"title": "TextGenerationPrefillToken",
+		"TextGenerationOutput": {
```
Comment (SBrandeis) [inline packages/tasks/src/tasks/text-generation/spec/output.json:23]: Same here - it seems this type is decalred twice (once at root level & once here in `$defs`)

**Review by SBrandeis — COMMENTED** [review]
This is very cool!!!
Very nice that it gets automatically generated from the OpenAPI spec

We are losing the `stream` parameter in the input types - not sure if it's a big deal though?

**Comment by SBrandeis** [comment]
Does this PR supersede #579?

File: packages/tasks/src/tasks/text-generation/inference.ts (hunk 1 above)
Comment (Wauplin) [reply packages/tasks/src/tasks/text-generation/inference.ts:39]: Documentation should be auto-generated from the TGI's specs. Currently `GenerateParameters` is not properly documented there which creates those deletions. I opened a PR in TGI to address this (https://github.com/huggingface/text-generation-inference/pull/1798). Once it's merged, we should be able to re-generate those without additional effort.

It is already the case for `chat_completion` parameters.

File: packages/tasks/src/tasks/text-generation/spec/input.json (hunk 4 above)
Comment (Wauplin) [reply packages/tasks/src/tasks/text-generation/spec/input.json:40]: Ahah, yeah^^
But actually it's used no-where at the moment. The Python code just ignore those rules.

**Comment by Wauplin** [comment]
> We are losing the stream parameter in the input types - not sure if it's a big deal though?

@SBrandeis  Fixed in https://github.com/huggingface/huggingface.js/pull/629/commits/97934afe33cc1739d3af57e524ff32d8ba3b4f96 by taking definition from `CompatGenerateRequest` instead of `GenerateRequest` (see TGI [code](https://github.com/huggingface/text-generation-inference/blob/5034f9ea8412323da3664ffdcd90be3f0c958718/router/src/lib.rs#L894)).

File: packages/tasks/src/tasks/text-generation/spec/input.json (hunk 3 above)
Comment (Wauplin) [reply packages/tasks/src/tasks/text-generation/spec/input.json]: Fixed in https://github.com/huggingface/huggingface.js/pull/629/commits/61b9d4e67f948239d3c4832ccabee866e2b67f91.

File: packages/tasks/src/tasks/text-generation/spec/output.json (hunk 5 above)
Comment (Wauplin) [reply packages/tasks/src/tasks/text-generation/spec/output.json]: Fixed in https://github.com/huggingface/huggingface.js/pull/629/commits/61b9d4e67f948239d3c4832ccabee866e2b67f91. Thanks for reporting :)

File: packages/tasks/src/tasks/text-generation/inference.ts (hunk 2 above)
Comment (Wauplin) [reply packages/tasks/src/tasks/text-generation/inference.ts:55]: Not sure what I can do about this unfortunately :confused:
It comes from

```js
"GrammarType": {
  "oneOf": [
    {
      "type": "object",
      "required": ["type", "value"],
      "properties": {
        "type": { "type": "string", "enum": ["json"] },
        "value": { "type": "string" }
      }
    },
    {
      "type": "object",
      "required": ["type", "value"],
      "properties": {
        "type": { "type": "string", "enum": ["regex"] },
        "value": { "type": "string" }
      }
    }
  ],
  "discriminator": { "propertyName": "type" }
}
``` 
in the original JSON specs. I tried playing with the import script but I feel a fix would have to live in `quicktype` itself (kinda annoying to fix/debug TBH)

**Comment by Wauplin** [comment]
@SBrandeis I addressed (almost) all comments. For the not-enough explicit `Type` type, I'm a bit lazy to go for a solution in guicktype directly. Maybe you'll have another idea to fix it more easily? We could have a ad-hoc rule but I'm not a big fan of it :grimacing: 

About docs for `text_generation` inputs, https://github.com/huggingface/text-generation-inference/pull/1798 has been merged but we'll have to wait for a release of TGI to get the updated openapi.json. In the meantime, I think it's fine not to have them documented and we'll fix that in a later PR (could even be automatized).

Otherwise, PR is almost good to review/merge I think. I still have the lint checks yelling at me but I don't have a good solution to deal with Any/unknown data without specify all attributes. Any suggestions? (I'd be fine disabling the rule for the entire file if that's possible TBH).

**Comment by radames** [comment]
@SBrandeis can @Wauplin go ahead and merge the last changes? this PR will be very helpful for a chat completion method we're working on thanks

**Comment by Wauplin** [comment]
Technically there is still the lint issue. Any advice? :pray: 

> PR is almost good to review/merge I think. I still have the lint checks yelling at me but I don't have a good solution to deal with Any/unknown data without specify all attributes. Any suggestions? (I'd be fine disabling the rule for the entire file if that's possible TBH).

**Comment by SBrandeis** [comment]
@Wauplin pushed a new version of the script that should pass the linter test - can you please double-check the logic is unchanged?

ping @coyotte508 for review as well - please let us know if you see anything that should not hit the codebase

**Comment by Wauplin** [comment]
Thanks for the changes @SBrandeis! Had a check to them and logic seems exactly the same yes :+1: 
I pushed https://github.com/huggingface/huggingface.js/pull/629/commits/52ec1e8ece0ca24aa2a2ecfeff3ee36093c1cee3 for prettier changes but that's all. Let's wait for a final review and then we should be able to merge :)

File: packages/tasks/package.json [hunk 6]
```diff
@@ -24,10 +24,11 @@
 		"format": "prettier --write .",
 		"format:check": "prettier --check .",
 		"prepublishOnly": "pnpm run build",
-		"build": "tsup src/index.ts --format cjs,esm --clean --dts && pnpm run inference-codegen",
+		"build": "tsup src/index.ts --format cjs,esm --clean --dts && pnpm run inference-tgi-import && pnpm run inference-codegen",
```
Comment (coyotte508) [inline packages/tasks/package.json:27]: Let's not run `inference-tgi-import` automatically.

It fetches a remote file, so let's run it once a while or in a cron job (to make a PR).

Fetching a remote file like this during a script that runs on `pnpm install` has some security implications, among other things

**Review by coyotte508 — COMMENTED** [review]
Besides my comment, LGTM

File: packages/tasks/package.json (hunk 6 above)
Comment (Wauplin) [reply packages/tasks/package.json]: Thanks for the review! Reverted that in https://github.com/huggingface/huggingface.js/pull/629/commits/092a7aa482c4f6e2a9a61596243692872b89e53e. About cron job/automated PRs, that's the plan yes but will be done in a future PR.

**Comment by Wauplin** [comment]
CI is green so time to merge! :smile: Thanks @SBrandeis @coyotte508 for the reviews.

@radames feel free to use those types in inference.js now

**Pull request merged**